src/commons/eventsourcing/store.rs

```diff
@@ -1,155 +1,69 @@
-    fn default() -> Self {
-        StoredValueInfo {
-            snapshot_version: 0,
-            last_event: 0,
-            last_command: 0,
-            last_update: Time::now(),
-        }
-    }
-}
-
-//------------ CommandKey ----------------------------------------------------
-
-#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
-pub struct CommandKey {
-    pub sequence: u64,
-    pub timestamp_secs: i64,
-    pub label: Label,
-}
-
-impl CommandKey {
-    pub fn new(sequence: u64, time: Time, label: Label) -> Self {
-        CommandKey {
-            sequence,
-            timestamp_secs: time.timestamp(),
-            label,
-        }
-    }
-
-    pub fn for_stored<S: WithStorableDetails>(command: &StoredCommand<S>) -> CommandKey {
-        CommandKey::new(command.sequence(), command.time(), command.details().summary().label)
-    }
-
-    pub fn matches_crit(&self, crit: &CommandHistoryCriteria) -> bool {
-        crit.matches_timestamp_secs(self.timestamp_secs)
-            && crit.matches_label(&self.label)
-            && crit.matches_sequence(self.sequence)
-    }
-}
-
-impl fmt::Display for CommandKey {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "command--{}--{}--{}", self.timestamp_secs, self.sequence, self.label)
-    }
-}
-
-impl FromStr for CommandKey {
-    type Err = CommandKeyError;
-
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        let parts: Vec<&str> = s.split("--").collect();
-        if parts.len() != 4 || parts[0] != "command" {
-            Err(CommandKeyError(s.to_string()))
-        } else {
-            let timestamp_secs = i64::from_str(parts[1]).map_err(|_| CommandKeyError(s.to_string()))?;
-            let sequence = u64::from_str(parts[2]).map_err(|_| CommandKeyError(s.to_string()))?;
-            // strip .json if present on the label part
-            let label = {
-                let end = parts[3].to_string();
-                let last = if end.ends_with(".json") {
-                    end.len() - 5
-                } else {
-                    end.len()
-                };
-                (end[0..last]).to_string()
-            };
-
-            Ok(CommandKey {
-                sequence,
-                timestamp_secs,
-                label,
-            })
-        }
-    }
-}
-
-//------------ CommandKeyError -----------------------------------------------
-
-#[derive(Clone, Debug, Eq, PartialEq)]
-pub struct CommandKeyError(String);
-
-impl fmt::Display for CommandKeyError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "invalid command key: {}", self.0)
-    }
-}
-
 //------------ AggregateStore ------------------------------------------------
 
 /// This type is responsible for managing aggregates.
 pub struct AggregateStore<A: Aggregate> {
     kv: KeyValueStore,
     cache: RwLock<HashMap<MyHandle, Arc<A>>>,
+    history_cache: Mutex<Option<HashMap<MyHandle, Vec<CommandHistoryRecord>>>>,
```

Comment:
I'm not sure if I'm thinking about this right but wouldn't that be better as `Option<Mutex>` ?

Reply:
Maybe... not sure. I changed it but then had to move things around a bit because of lifetimes. It will be slightly more efficient in the non-default case where no cache is used.